From bf3ca9bb6a070d1530d4bbddfb1afe729b7df9db Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 13:12:40 -0400
Subject: [ethernet] i40evf: resend FW request if no response

Message-id: <1409663591-32531-118-git-send-email-sassmann@redhat.com>
Patchwork-id: 90590
O-Subject: [RHEL7.1 PATCH 117/148] i40evf: resend FW request if no response
Bugzilla: 1091127
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

Sometimes the firmware will not indicate an error but fail to pass a
message between the VF and the PF driver. If this happens, just resend
the request.

This fixes an initialization failure if many VFs are instantiated at the
same time and the VF module is autoloaded.

Change-ID: Idd1ad8da2fd5137859244685c355941427d317d7
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 56f9920a952130a7cd20952b3525bc238a87e489)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40evf/i40evf_main.c b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
index 7138ab4..8082a9f 100644
--- a/drivers/net/ethernet/intel/i40evf/i40evf_main.c
+++ b/drivers/net/ethernet/intel/i40evf/i40evf_main.c
@@ -2018,6 +2018,10 @@ static void i40evf_init_task(struct work_struct *work)
   if (err) {
    dev_info(&pdev->dev, "Unable to verify API version (%d), retrying\n",
     err);
+   if (err == I40E_ERR_ADMIN_QUEUE_NO_WORK) {
+    dev_info(&pdev->dev, "Resending request\n");
+    err = i40evf_send_api_ver(adapter);
+   }
    goto err;
   }
   err = i40evf_send_vf_config_msg(adapter);
-- 
1.7.1